File: scratch/15.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # we want all tuples of 3 integers such that none of them are duplicates
        # and those 3 integers sum to be 0
        solutions = []
        nums.sort()

        # the reason we loop until -2 of the size is because we're summing 3 values
        for i in range(len(nums)-2):
            # i = 0 cannot be a duplicate, otherwise check i and prev for duplicate
            if (i > 0 and (nums[i] == nums[i - 1])):
                continue

            # create left and right pointers
            l, r = i + 1, len(nums) - 1

            # while left and right have not crossed
            while l < r:
                # take the sum
                s = nums[i] + nums[l] + nums[r]
                # if sum is neg, increment left
                if s < 0:
                    l += 1
                # if sum is pos, decrement right
                elif s > 0:
                    r -= 1
                # otherwise it's zero so append the solution onto the results
                else:
                    solutions.append([nums[i], nums[l], nums[r]])

                    # then do this for some reason
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                    l += 1; r -= 1

        return solutions

s = Solution()
print(s.threeSum([1, 2, 3, 4, -5]))

File: scratch/15.cpp
#include <bits/stdc++.h>
#define N size
#define pb push_back
#define l int
#define vi vector<l>

using namespace std;

vector<vi> threeSum(vi ns) {
	vector<vi> res;
	sort(ns.begin(), ns.end());

	for (l i = 0; i < ns.N() - 2; i++) {
		// check for duplicates
		if (i > 0 && ns[i] == ns[i - 1]) continue;

		l lp = i + 1, rp = ns.N() - 1;
		while (lp < rp) {
			// the sum of these 3 values
			l sum = ns[i] + ns[lp] + ns[rp];
			if (sum > 0) {
				// if pos we want to dec the value
				// nudging it towards zero
				rp--;
			}
			else if (sum < 0) {
				// if neg, then we want to inc the value
				lp++;
			}
			else {
				// push back in the case that it's zero
				res.pb({ns[i], ns[lp], ns[rp]});

				// duplicate checking
				while (lp < rp && ns[lp] == ns[lp-1]) lp++;
				while (lp < rp && ns[rp] == ns[rp-1]) rp--;
				lp++; rp--;
			}
		}
	}
	return res;
}

int main() {
}
